juju/bundle.py

```diff
@@ -156,311 +171,663 @@ def resolve(self, reference):
+        await context.client_facade.AddCharm(channel=None, url=entity_id, force=False)
         return entity_id
 
-    async def addMachines(self, change_id, params=None):
-        """
-        :param params dict:
-            Dictionary specifying the machine to add. All keys are optional.
-            Keys include:
+    def __str__(self):
+        series = ""
+        channel = ""
+        if self.series != "":
+            series = " for series {}".format(self.series)
+        if self.channel is not None:
+            channel = " from channel {}".format(self.channel)
+        return "upload charm {charm}{series}{channel}".format(charm=self.charm,
+                                                              series=series,
+                                                              channel=channel)
 
-            series: string specifying the machine OS series.
 
-            constraints: string holding machine constraints, if any. We'll
-                parse this into the json friendly dict that the juju api
-                expects.
+class AddMachineChange(ChangeInfo):
+    _toPy = {'series': 'series',
+             'constraints': 'constraints',
+             'container-type': 'container_type',
+             'parent-id': 'parent_id'}
 
-            container_type: string holding the type of the container (for
-                instance ""lxd" or kvm"). It is not specified for top level
-                machines.
+    """
+    AddMachineChange holds a change for adding a machine or container.
 
-            parent_id: string holding a placeholder pointing to another
-                machine change or to a unit change. This value is only
-                specified in the case this machine is a container, in
-                which case also ContainerType is set.
+    :series: holds the optional machine OS series.
+    :constraints: holds the optional machine constraints.
+    :container_type: optionally holds the type of the container (for instance
+        "lxc" or kvm"). It is not specified for top level machines.
+    :parent_id: holds the id of the parent machine.
+    """
+    def __init__(self, change_id, requires, params=None):
+        super(AddMachineChange, self).__init__(change_id, requires)
+        # this one is weird, as it returns a set of parameters inside a list.
+        if isinstance(params, list):
+            options = params[0] or {}
+            self.series = options.get("series")
+            self.constraints = options.get("constraints")
+            self.container_type = options.get("containerType")
+            self.parent_id = options.get("parentId")
+        elif isinstance(params, dict):
+            AddMachineChange.from_dict(self, params)
+        else:
+            raise Exception("unexpected params type")
 
+    @staticmethod
+    def method():
         """
-        params = params or {}
+        method returns an associated ID for the Juju API call.
```

Comment:
Convention is that the docstring starts after the """ vs on a newline:

https://www.python.org/dev/peps/pep-0257/#id17